File: promise/src/test/java/com/voxeet/testpromise/resolve/PromiseTestSimpleResolve.java
package com.voxeet.testpromise.resolve;

import com.voxeet.promise.Promise;
import com.voxeet.promise.solve.PromiseSolver;
import com.voxeet.testpromise.utils.AndroidMockUtil;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class PromiseTestSimpleResolve {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);

        final boolean[] catched = {true};
        System.out.println("executing test");
        new Promise<>((PromiseSolver<String>) solver -> {
            catched[0] = false;
            solver.resolve("called");
            latch.countDown();
        }).execute();

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (catched[0]) {
            throw new IllegalStateException("Expected 0 error...");
        }
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/resolve/PromiseTestSimpleResolve.kt
package com.voxeet.testpromise.resolve

import com.voxeet.promise.HandlerFactory
import com.voxeet.promise.Promise
import com.voxeet.promise.await
import com.voxeet.promise.solve.Solver
import com.voxeet.promise.solve.ThenVoid
import com.voxeet.testpromise.mockedhandler
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import org.junit.Before
import org.junit.Test
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue

@OptIn(ExperimentalCoroutinesApi::class)
class PromiseTestSimpleResolve {
    @Before
    fun setHandler() {
        HandlerFactory.setHandler(mockedhandler())
    }

    @Test
    fun test() = runTest {
        var called = false

        Promise { solver: Solver<String?> ->
            called = true

            solver.resolve(null)
        }.await()

        assertTrue(called)
    }

    @Test
    fun `test then void`() = runTest {
        var called = false

        val void = Promise { solver: Solver<String?> -> solver.resolve(null) }
            .then { it }
            .then(ThenVoid {
                called = true
            }).await()

        assertTrue(called)
        assertNull(void)
    }
}
